quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+                                expected: *scalar_type,
+                                found: memory_region.size.data_type,
+                            });
+                        }
+                        Ok(memory_region.size.clone())
+                    }
+                }?;
+                Ok(ResolvedCallArgument::Vector {
+                    memory_region_name: value.clone(),
+                    vector: expected_vector,
+                    mutable: extern_parameter.mutable,
+                })
+            }
+            UnresolvedCallArgument::MemoryReference(value) => {
+                let expected_scalar = match extern_parameter.data_type {
+                    ExternParameterType::Scalar(ref scalar) => Ok(scalar),
+                    ExternParameterType::FixedLengthVector(_)
+                    | ExternParameterType::VariableLengthVector(_) => {
+                        Err(CallArgumentResolutionError::InvalidVectorArgument(
+                            Self::MemoryReference(value.clone()),
+                        ))
+                    }
+                }?;
+                let memory_region = memory_regions.get(value.name.as_str()).ok_or_else(|| {
+                    CallArgumentResolutionError::UndeclaredMemoryReference(value.name.clone())
+                })?;
+                if memory_region.size.data_type != *expected_scalar {
+                    return Err(CallArgumentResolutionError::MismatchedScalar {
+                        expected: *expected_scalar,
+                        found: memory_region.size.data_type,
+                    });
+                }
+                Ok(ResolvedCallArgument::MemoryReference {
+                    memory_reference: value.clone(),
+                    scalar_type: *expected_scalar,
+                    mutable: extern_parameter.mutable,
+                })
+            }
+            UnresolvedCallArgument::Immediate(value) => {
+                let expected_scalar = match extern_parameter.data_type {
+                    ExternParameterType::Scalar(ref scalar) => Ok(scalar),
+                    ExternParameterType::FixedLengthVector(_)
+                    | ExternParameterType::VariableLengthVector(_) => Err(
+                        CallArgumentResolutionError::InvalidVectorArgument(self.clone()),
+                    ),
+                }?;
+                Ok(ResolvedCallArgument::Immediate {
+                    value: *value,
+                    scalar_type: *expected_scalar,
+                })
+            }
+        }
+    }
+
+    /// Check if the argument is compatible with the return type of the [`ExternSignature`]. If so,
+    /// return the appropriate [`ResolvedCallArgument`]. If not, return an error.
+    fn resolve_return(
+        &self,
+        memory_regions: &IndexMap<String, MemoryRegion>,
+        return_type: ScalarType,
+    ) -> Result<ResolvedCallArgument, CallArgumentResolutionError> {
```

Comment:
The return type of this function is misleading – it's not a call argument.  Can we rename the type, and maybe share code with the above function?